### PR TITLE
Bugfix/teamred 31

### PR DESCRIPTION
**Pull request by sasa-bonta**
Fixed bugs from precedent sprints. Added some new features that the green team has implemented earlier. This is a very important ticket!!!!!!!!! New features should start from this ticket. Please verify very careful if it doesn't have bugs!!!

### CODE REVIEW

File: .idea/php.xml [hunk 1]
```diff
@@ -107,6 +107,7 @@
       <path value="$PROJECT_DIR$/vendor/symfony/webpack-encore-bundle" />
       <path value="$PROJECT_DIR$/vendor/doctrine/doctrine-fixtures-bundle" />
       <path value="$PROJECT_DIR$/vendor/doctrine/data-fixtures" />
+      <path value="$PROJECT_DIR$/vendor/symfonycasts/verify-email-bundle" />
```
Comment (aharabara) [inline .idea/php.xml]: create a ticket to remove `.idea/` from git repository

File: migrations/Version20210505134027.php [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+{
+    public function getDescription() : string
+    {
+        return '';
+    }
+
+    public function up(Schema $schema) : void
+    {
+        // this up() migration is auto-generated, please modify it to your needs
+        $this->addSql('ALTER TABLE user ADD activated TINYINT(1) DEFAULT \'0\' NOT NULL');
```
Comment (aharabara) [inline migrations/Version20210505134027.php]: previous migration contains same altering. Why do you need 2 migrations that are performing same actions?

File: migrations/Version20210505134032.php [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+{
+    public function getDescription() : string
+    {
+        return '';
+    }
+
+    public function up(Schema $schema) : void
+    {
+        // this up() migration is auto-generated, please modify it to your needs
+        $this->addSql('ALTER TABLE user ADD activated TINYINT(1) DEFAULT \'0\' NOT NULL');
```
Comment (aharabara) [inline migrations/Version20210505134032.php:23]: Same here, duplicated functionality

File: src/Controller/AdminController.php [hunk 4]
```diff
@@ -55,29 +63,35 @@ public function list(ProductRepository $productRepository, Request $request): Re
...
     {
         $product = new Product();
         $form = $this->createForm(ProductType::class, $product);
         $form->handleRequest($request);
-
+        $errors = [];
+        if ($form->isSubmitted() && !$form->isValid()) {
+            $repo = $this->getDoctrine()->getRepository(Product::class);
+            if ($repo->count(['code' => $product->getCode()]) > 0) {
+                array_push($errors, "This code already exists");
```
Comment (aharabara) [inline src/Controller/AdminController.php]: you can use `$this->addFlash()` for such cases?

File: src/Controller/AdminController.php [hunk 5]
```diff
@@ -55,29 +63,35 @@ public function list(ProductRepository $productRepository, Request $request): Re
...
+                ]);
+            }
+        }
         if ($form->isSubmitted() && $form->isValid()) {
-
-            # Errors existent code
-            $errors = [];
             $repo = $this->getDoctrine()->getRepository(Product::class);
             if ($repo->count(['code' => $product->getCode()]) > 0) {
                 array_push($errors, "This code already exists");
```
Comment (aharabara) [inline src/Controller/AdminController.php]: same, `addFlash`

File: src/Controller/ProductApiController.php [hunk 6]
```diff
@@ -26,18 +25,27 @@
  */
 class ProductApiController extends AbstractController
 {
+    private function makeResponse400($message): JsonResponse
+    {
+        $response = new JsonResponse();
```
Comment (aharabara) [inline src/Controller/ProductApiController.php]: You can create a `BadRequestJsonResponse` class instead of this method, but in general you can throw BadREquestException and specify endpoint `_format` as `json`, so it will cast exception into a json message

File: src/Controller/ProductApiController.php [hunk 7]
```diff
@@ -47,63 +55,101 @@ public function index(ProductRepository $productRepository, Request $request): R
         try {
             $searchCriteria = new SearchCriteria($name, $category, $page, $limit, $order, $ascDesc);
         } catch (InvalidPageException $e) {
```
Comment (aharabara) [inline src/Controller/ProductApiController.php]: use one catch with exceptions parent class and throw a `BadRequestException`. If you have questions about this case - ping me

File: src/Controller/ProductApiController.php [hunk 8]
```diff
@@ -47,63 +55,101 @@ public function index(ProductRepository $productRepository, Request $request): R
...
-    public function new(Request $request): JsonResponse
+    public function new(Request $request, ProductRepository $repo): JsonResponse
     {
+        $response = new JsonResponse();
         $parameters = json_decode($request->getContent(), true);
         $product = new Product();
         $entityManager = $this->getDoctrine()->getManager();
         $dateTime = new DateTime(null, new DateTimeZone('Europe/Athens'));
+
+        if (empty($parameters['code']) or empty($parameters['name']) or empty($parameters['category']) or empty($parameters['price']) or empty($parameters['description'])) {
```
Comment (aharabara) [inline src/Controller/ProductApiController.php]: 1. Don't use `or` - it is a little bit dangerous, it has a different order of execution
2. Use product form class to validate data instead of validating it manually

File: src/Controller/ProductApiController.php [hunk 9]
```diff
@@ -47,63 +55,101 @@ public function index(ProductRepository $productRepository, Request $request): R
...
      */
-    public function edit(Request $request, Product $product, ProductRepository $productRepository): JsonResponse
+    public function edit(Request $request, Product $product, ProductRepository $repo): JsonResponse
     {
+        $response = new JsonResponse();
         $parameters = json_decode($request->getContent(), true);
-        $product = $productRepository->findOneBy(array('code' => $parameters['code']));
+
+        if ($repo->findOneBy(array('code' => $parameters['code'])) === null) {
+            $response->setStatusCode(JsonResponse::HTTP_NO_CONTENT);
```
Comment (aharabara) [inline src/Controller/ProductApiController.php]: `NO_CONTENT` response should be empty

File: src/Controller/ProductApiController.php [hunk 10]
```diff
@@ -47,63 +55,101 @@ public function index(ProductRepository $productRepository, Request $request): R
...
+            $data = ["status" => 204, "description" => "no content","message" => "the product with this code doesn't exist"];
+            $response->setData($data);
+            return $response;
+        }
+
         $entityManager = $this->getDoctrine()->getManager();
         $dateTime = new DateTime(null, new DateTimeZone('Europe/Athens'));
+
+        if (empty($parameters['code']) or empty($parameters['name']) or empty($parameters['category']) or empty($parameters['price']) or empty($parameters['description'])) {
```
Comment (aharabara) [inline src/Controller/ProductApiController.php]: same as in `::new` - use forms

File: src/Controller/UserController.php [hunk 11]
```diff
@@ -60,6 +56,20 @@ public function index(Request $request, UserRepository $userRepository): Respons
         ]);
     }
 
+    private function checkDataNewUser(User $user): array
```
Comment (aharabara) [inline src/Controller/UserController.php]: Looks like you have more and more duplicated code. You need to move it to a different service, but leave it for a different PR. You can create a ticket for this

File: src/Form/RegisterType.php [hunk 12]
```diff
@@ -17,12 +17,7 @@ class RegisterType extends AbstractType
     public function buildForm(FormBuilderInterface $builder, array $options)
     {
         $builder
-            ->add('username', TextType::class, [
-                'constraints' => new Length([
```
Comment (aharabara) [inline src/Form/RegisterType.php:21]: why did you remove this constraint? I can pass now a very long string and it will fail when database insert will happen
By the way - try to move your validation checks to forms

**Review by OsipciucIon19 — APPROVED** [review]
Error handling for user entity now works fine. 

The bug with “-“ between [show] and [delete] buttons was deleted.

Errors are displayed correctly.

Search is implemented correctly.

Errors for apis are now correctly implemented.

The path /admin redirects to /admin/products.

Email confirmation works.

The name input is saved when searching.

Limit and sort by now corresponds with the url.

Categories are displayed correctly.

 

Great job!

File: templates/admin/navbar_admin_products.html.twig [hunk 13]
```diff
@@ -85,44 +85,54 @@
...
     }
 
     function setName() {
         var inputVal = document.getElementById("myInput").value;
-        var urlSearchParams = new URL('http://localhost:8000/admin/products/');
-        urlSearchParams.searchParams.set('name', inputVal);
-        urlSearchParams.searchParams.set('page', 1);
-        window.location.replace(urlSearchParams);
+        var urlSearchParams;
+        if (window.location.origin + window.location.pathname !== "http://localhost:8000/admin/products/") {
```
Comment (aharabara) [inline templates/admin/navbar_admin_products.html.twig:126]: dont use absolute URL - if I will try to run this project on a different machine with a proper url - it will not work

File: templates/admin/user/index.html.twig [hunk 14]
```diff
@@ -50,6 +52,12 @@
                             <td>{{ user.id }}</td>
                             <td>{{ user.username }}</td>
                             <td>{{ user.email }}</td>
+                            <td>{% if user.activated == 1 %}
```
Comment (aharabara) [inline templates/admin/user/index.html.twig]: not sure, but I think you can use `{{ user.activated ? 'true' : 'false' }}`

File: templates/admin/user/index.html.twig [hunk 15]
```diff
@@ -59,9 +67,11 @@
                                 <a href="{{ path('user_show', {'id': user.id}) }}">
                                     <button class="btn btn-success">Show</button>
                                 </a>
-                                <a href="{{ path('user_edit', {'id': user.id}) }}">
-                                    <button class="btn btn-warning">Edit</button>
-                                </a>
+                                {% if user.username != 'sasa' %}
```
Comment (aharabara) [inline templates/admin/user/index.html.twig:66]: hmmmm, do you really need this line? or maybe you want to have role check

File: templates/admin/user/show.html.twig [hunk 16]
```diff
@@ -38,11 +47,12 @@
...
     </p>
     <br>
-    <p align="center">
-        <a href="{{ path('user_edit', {'id': user.id}) }}">
-            <button class="btn btn-success">Edit</button>
-        </a>
-    </p>
-
-    {{ include('admin/user/_delete_form.html.twig') }}
+    {% if user.username != 'sasa' %}
```
Comment (aharabara) [inline templates/admin/user/show.html.twig:46]: Same here - I suppose you need a role check

**Review by aharabara — CHANGES_REQUESTED** [review]

File: .idea/php.xml (hunk 1 above)
Comment (sasa-bonta) [reply .idea/php.xml]: Maybe next time :D

File: migrations/Version20210505134027.php (hunk 2 above)
Comment (sasa-bonta) [reply migrations/Version20210505134027.php]: Fixed

File: migrations/Version20210505134032.php (hunk 3 above)
Comment (sasa-bonta) [reply migrations/Version20210505134032.php:23]: Fixed

File: src/Controller/AdminController.php (hunk 4 above)
Comment (sasa-bonta) [reply src/Controller/AdminController.php]: Added flashes. The code is a little shorter. But here you haven't noticed duplication of $repo = $this->getDoctrine()->getRepository(Product::class);. I've moved it higher and duplication is eliminated

File: src/Controller/AdminController.php (hunk 5 above)
Comment (sasa-bonta) [reply src/Controller/AdminController.php]: done

File: src/Controller/ProductApiController.php (hunk 6 above)
Comment (sasa-bonta) [reply src/Controller/ProductApiController.php]: Good idea. Maybe next time. But at the moment my solution works fine and shows correspondent error message in json

File: src/Controller/ProductApiController.php (hunk 8 above)
Comment (sasa-bonta) [reply src/Controller/ProductApiController.php]: I've created ProductForm class and there I verify if each parameter is present, else throw Exception.  Line 86 instead of if try new ProductForm and catch Exception

File: templates/admin/user/show.html.twig (hunk 16 above)
Comment (sasa-bonta) [reply templates/admin/user/show.html.twig:46]: I made it to prevent deleting of my admin user

File: src/Form/RegisterType.php (hunk 12 above)
Comment (sasa-bonta) [reply src/Form/RegisterType.php:21]: Redo

File: templates/admin/navbar_admin_products.html.twig (hunk 13 above)
Comment (sasa-bonta) [reply templates/admin/navbar_admin_products.html.twig:126]: Here I'll think later

File: templates/admin/user/index.html.twig (hunk 14 above)
Comment (sasa-bonta) [reply templates/admin/user/index.html.twig]: {{ user.activated ? 'true' : 'false' }} Worked perfectly

File: templates/admin/user/show.html.twig (hunk 16 above)
Comment (aharabara) [reply templates/admin/user/show.html.twig:46]: You can use such thing as privelege and as possibility to have account that can't be deleted. Your can create a ticket for that later

**Review by aharabara — APPROVED** [review]
